code/processes/tailsort.q

```diff
@@ -0,0 +1,77 @@
+\d .ts
+
+taildir:hsym `$getenv`KDBTAIL;                                             /-load in taildir env variables
+hdbdir:hsym `$getenv`KDBHDB;                                               /-load in hdb env variables
+savelist:@[value;`savelist;`quote`trade];                                  /-list of tables to save to HDB
+taildbs:key taildir;                                                       /-list of tailDBs that need saved to HDB
+taildirs:();                                                               /-empty list to append tailDB paths to - to be used
+                                                                           / when HDB save is complete to delete tailDB partitions
+\d .
+
+savescompleted:0;                                                          /-variable to count how many tailDBs have been saved to HDB
+
+endofday:{[pt;procname]
+  /- function to trigger data load & save to HDB once endofday message is received from tailer(s)
+  .lg.o[`endofday;"end of day message received from ",string[procname]," - ",string[pt]];
+  loadandsave[pt;procname];
+  };
+
+loadandsave:{[pt;procname]
+  /-function to merge tables from subpartitions in tailDB and save to HDB
+  taildir:` sv (.ts.taildir;procname;`$string pt);
+  .ts.taildirs,:taildir;
+  /-merge tables from tailDBs and save to HDB
+  mergebypart[taildir;pt;;.ts.hdbdir] each .ts.savelist;
+  /-increase savescompleted counter
+  savescompleted+::1;
+  .lg.o[`sortcomplete;"end of day sort complete for ",string[procname]];
+  /-check if all eod saves have been completed, if so trigger savecomplete
+  if[savescompleted = count .ts.taildbs;savecomplete[pt;.ts.savelist]];
+  };
+
+mergebypart:{[dir;pt;tabname;dest]
+  /-function to merge table partitions from tailDB and save to HDB
+  /-get list of partitions to be merged
+  parts:(key dir) except `access;
+  partdirs:{` sv (x;y;z)}[dir;;tabname] each parts;
+  /-load data from each partition
+  .lg.o[`merge;"reading partition(s) ", (", " sv string[partdirs])];
+  data:get each partdirs;
+  /-if multiple partitions have been read in data will be a list of tabs, if this is the case - join into single tab
+  if[98<>type data;data:(,/)data];
+  /-upsert data to partition in destination directory
+  dest:` sv .Q.par[dest;pt;tabname],`;
+  .[upsert;
+    (dest;data);
+    {[e] .lg.e[`upserttopartition;"failed to save table to disk : ",e];'e}
```

Comment:
is it intended that the error is signalled here? Or do we want to log the error and continue?

Reply:
I put the error here as I thought this was convention based on other merge functions/upserttopartition functions

Reply:
I think we need to decide on the intended behaviour here - signalling an error will mean execution is interrupted and nothing after this will execute - so in `loadandsave` we'll never update the `savescompleted` counter etc. - is that what we want to happen?

Reply:
> but I guess the issue with signalling the error would be if it happened eg during the save for the first tailer and so would prevent any saves from tailers waiting to be sorted?

I don't think it should prevent that, as those will each be triggered by separate IPC messages I think? the signal will only interrupt the excution of the current IPC message

Reply:
okay yeah, then I think we should signal the error to ensure that we don't lose the data that failed to save